pkg/config/instrumentation/config_metrics.go

```diff
@@ -0,0 +1,59 @@
+package instrumentation
+
+import (
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type configMetrics struct {
+	configHash        *prometheus.GaugeVec
+	configLoadSuccess prometheus.Gauge
+	configLoadSeconds *prometheus.GaugeVec
+}
+
+// ConfigMetrics exposes metrics related to configuration loading
+var ConfigMetrics = newConfigMetrics()
+
+func newConfigMetrics() *configMetrics {
+	var m configMetrics
+
+	m.configHash = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "agent_config_hash",
+			Help: "Hash of the currently active config file.",
+		},
+		[]string{"sha256"},
+	)
+	m.configLoadSuccess = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "agent_config_last_load_successful",
+		Help: "Config loaded successfully.",
+	})
+	m.configLoadSeconds = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "agent_config_last_load_timestamp_seconds",
+		Help: "Timestamp of the last configuration load by result.",
+	}, []string{"result"})
+
+	return &m
+}
+
+// Create a sha256 hash of the config before expansion and expose it via
+// the agent_config_hash metric.
+func (c *configMetrics) InstrumentConfig(buf []byte) {
+	hash := sha256.Sum256(buf)
+	c.configHash.Reset()
+	c.configHash.WithLabelValues(fmt.Sprintf("%x", hash)).Set(1)
+}
+
+// Expose metrics for load success / failures.
+func (c *configMetrics) InstrumentLoad(success bool) {
```

Comment:
nit: Passing success to a function like this feels a little weird to me. You could implement this using some middleware-y function:

```go
cm.InstrumentLoad(bb, func(bb []byte) error {
  // do the load 
  // return if there was an error 
})
```

which would set all the appropriate metrics for you. 

Honestly, I'm not sure if that's really worth implementing right now. This is mostly just a nit, it feels a little strange but I wouldn't say it's a blocker.

Reply:
I agree with your point here, however looking at the config code, it is hard to agree on a single API that would work for all current implementations (including dynamic config). I am also purposely trying to separate config hash instrumentation from config load (under the assumption that it is better to build the hash before applying dynamic transformations like variable substitution, etc to prevent noise, while it is better to instrument loads as late as possible, to cover all possible failure modes). I am also open to have these assumptions challenged :)

Reply:
I don't feel too strongly about it, it's something we can always change later and it's not intended to be a public facing API. It's more than fine for now 👍